File: code/216.combination-sum-iii.php
<?php
/*
 * @lc app=leetcode id=216 lang=php
 *
 * [216] Combination Sum III
 */

// @lc code=start
class Solution {

    /**
     * @param Integer $k
     * @param Integer $n
     * @return Integer[][]
     */
    function combinationSum3($k, $n) {
        return $this->recursive($k, $n, 1);
    }

    function recursive($k, $n, $start) {
        if (array_sum(range($start, $start + $k - 1)) > $n) {
            return [];
        }

        if ($k == 1) {
            if ($n >= $start && $n <= 9) {
                return [[$n]];
            } else {
                return [];
            }
        }

        $result = [];
        for ($i = $start; $i < 10; $i++) {
            $combinations = $this->recursive($k - 1, $n - $i, $i + 1);

            if ($combinations) {
                foreach ($combinations as $combination) {
                    $result[] = [
                        $i,
                        ...$combination,
                    ];
                }
            }
        }

        return $result;
    }
}
// @lc code=end

File: code/216.combination-sum-iii.go
package code

/*
 * @lc app=leetcode id=216 lang=golang
 *
 * [216] Combination Sum III
 */

// @lc code=start
func combinationSum3(k int, n int) [][]int {
	ans := make([][]int, 0)

	var backtracking func(combination []int, from int, remain int)
	backtracking = func(combination []int, from int, remain int) {
		if len(combination) == k && remain == 0 {
			// コピーしないと以前のスライスの中身が上書きされてしまう
			clone := make([]int, k)
			copy(clone, combination)
			ans = append(ans, clone)
			return
		}

		if len(combination) >= k || remain <= 0 {
			return
		}

		for i := from; i < 10; i++ {
			new := append(combination, i)
			backtracking(new, i+1, remain-i)
		}
	}

	backtracking([]int{}, 1, n)
	return ans
}

// @lc code=end
